### PR TITLE
Support Cast In Eval Engine

### PR DESCRIPTION
**Pull request by yliuuuu**
## Relevant Issues
- [Closes/Related To] Issue #XXX

## Description
In this PR, we model Cast as its own Plan Node. 

Added Plan Node: 
```
cast::{
  operand: partiql_value_type, // operand is a value.
  target: partiql_value_type, // TODO: Consider change this to Static Type
  cast_type: cast_type::[
    COERCION,
    EXPLICIT,
    UNSAFE
  ],
}

rex::{
  op:[
    cast_op::{
      arg: rex,
      cast: cast
    },
  ]
}    
```

Internal Plan Node is modeled exactly the same as public Plan Node for now. 

### Cast Table
The cast table are hard coded for now. Check [here](https://github.com/partiql/partiql-lang-kotlin/blob/2d493c625e09c2a0695a707e6b41841b9b73e2d6/partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/TypeCasts.kt) for the cast table. 

### Planning
During planning, the planner will check if there is a available cast relationship exist. If so, the planner store the cast relationship in cast Node, and the `castOp` node points to the cast Node. 

Note that if in planning time, we know that the arg is NULL, we reduce the cast op to a `RexOpLit($typedNullValue)` with type as the target type. If the arg is `missing` then we reduce the cast op to a `RexOpLit(missingValue())`

Note that for dynamic function, instead of storing a function signature that hints on the correct cast to invoke, we store a Cast ref to achieve the same functionality. 

Example: 
```
t_int2 : StaticType.INT2
t2: StaticType.unionOf(INT4, DECIMAL) 

t_int2 + t2 

⚬ PartiQLPlan[tag=Plan-9835b5f]
   ├──Catalog[name=test, tag=Plan-b2e2597c]
   │  ├──Symbol[type=int2, tag=Plan-90f65f9e]
   │  └──Symbol[type=union(decimal, int4), tag=Plan-1ba87674]
   └──Query[tag=Plan-810bfdb7]
      └──Rex[type=union(decimal, int4), tag=Plan-df596342]
         └──Dynamic[tag=Plan-8aa94980]
            ├──Rex[type=int2, tag=Plan-3ff7f113]
            │  └──Global[tag=Plan-7675096]
            │     └──Ref[catalog=0, symbol=0, tag=Plan-7412482]
            ├──Rex[type=union(decimal, int4), tag=Plan-c84905a8]
            │  └──Global[tag=Plan-ec8cbafe]
            │     └──Ref[catalog=0, symbol=1, tag=Plan-443bf140]
            ├──Candidate[tag=Plan-8502a68c]
            │  ├──Fn[signature=PLUS__DECIMAL_ARBITRARY_DECIMAL_ARBITRARY__DECIMAL_ARBITRARY, tag=Plan-89b81583]
            │  └──Cast[castType=COERCION, operand=INT16, target=DECIMAL_ARBITRARY, tag=Plan-530ad5f8]
            └──Candidate[tag=Plan-324b0bb7]
               ├──Fn[signature=PLUS__INT32_INT32__INT32, tag=Plan-8ced4d96]
               └──Cast[castType=COERCION, operand=INT16, target=INT32, tag=Plan-cfbb004f]
```

### Updated Cast Table
|ANY| BOOL| INT8| INT16| INT32| INT64| INT| DECIMAL| DECIMAL_ARBITRARY| FLOAT32| FLOAT64| CHAR| STRING| SYMBOL| BINARY| BYTE| BLOB| CLOB| DATE| TIME| TIMESTAMP| INTERVAL| BAG| LIST| SEXP| STRUCT| NULL| MISSING|
|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|
| ANY | ⬤ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ |
| BOOL | X | ⬤ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| INT8 | X | ◯ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ◯ | ⬤ | ⬤ | ⬤ | X | ◯ | ◯ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| INT16 | X | ◯ | △ | ⬤ | ⬤ | ⬤ | ⬤ | ◯ | ⬤ | ⬤ | ⬤ | X | ◯ | ◯ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| INT32 | X | ◯ | △ | △ | ⬤ | ⬤ | ⬤ | ◯ | ⬤ | ⬤ | ⬤ | X | ◯ | ◯ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| INT64 | X | ◯ | △ | △ | △ | ⬤ | ⬤ | ◯ | ⬤ | ⬤ | ⬤ | X | ◯ | ◯ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| INT | X | ◯ | △ | △ | △ | △ | ⬤ | ◯ | ⬤ | ⬤ | ⬤ | X | ◯ | ◯ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| DECIMAL | X | ◯ | ◯ | ◯ | ◯ | ◯ | X | ◯ | ◯ | ◯ | ◯ | X | ◯ | ◯ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| DECIMAL_ARBITRARY | X | ◯ | △ | △ | △ | △ | ◯ | ◯ | ⬤ | ◯ | ◯ | X | ◯ | ◯ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| FLOAT32 | X | ◯ | △ | △ | △ | △ | ◯ | ◯ | ⬤ | ⬤ | ⬤ | X | ◯ | ◯ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| FLOAT64 | X | ◯ | △ | △ | △ | △ | ◯ | ◯ | ⬤ | X | ⬤ | X | ◯ | ◯ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| CHAR | X | △ | X | X | X | X | X | X | X | X | X | ⬤ | ⬤ | ⬤ | X | X | X | X | X | X | X | X | X | X | X | X | X | X |
| STRING | X | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | X | ⬤ | ◯ | X | X | X | ⬤ | X | X | X | X | X | X | X | X | X | X |
| SYMBOL | X | △ | △ | △ | △ | △ | △ | △ | △ | △ | △ | X | ⬤ | ⬤ | X | X | X | ⬤ | X | X | X | X | X | X | X | X | X | X |
| BINARY | X | X | X | X | X | X | X | X | X | X | X | X | X | X | ⬤ | X | X | X | X | X | X | X | X | X | X | X | X | X |
| BYTE | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | ⬤ | X | X | X | X | X | X | X | X | X | X | X | X |
| BLOB | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | ⬤ | X | X | X | X | X | X | X | X | X | X | X |
| CLOB | X | X | X | X | X | X | X | X | X | X | X | △ | △ | △ | X | X | X | ⬤ | X | X | X | X | X | X | X | X | X | X |
| DATE | X | X | X | X | X | X | X | X | X | X | X | ◯ | ◯ | ◯ | X | X | X | X | ⬤ | X | ◯ | X | X | X | X | X | X | X |
| TIME | X | X | X | X | X | X | X | X | X | X | X | ◯ | ◯ | ◯ | X | X | X | X | X | ⬤ | ◯ | X | X | X | X | X | X | X |
| TIMESTAMP | X | X | X | X | X | X | X | X | X | X | X | ◯ | ◯ | ◯ | X | X | X | X | ◯ | ◯ | ⬤ | X | X | X | X | X | X | X |
| INTERVAL | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | ⬤ | X | X | X | X | X | X |
| BAG | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | ⬤ | ◯ | ◯ | X | X | X |
| LIST | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | ◯ | ⬤ | ◯ | X | X | X |
| SEXP | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | ◯ | ◯ | ⬤ | X | X | X |
| STRUCT | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | ⬤ | X | X |
| NULL | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | ⬤ | X |
| MISSING | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X | X |

## Other Information
- Updated Unreleased Section in CHANGELOG: **[YES/NO]**
  - < If NO, why? >

- Any backward-incompatible changes? **[YES/NO]**
  - < If YES, why? >
  - < For this purpose, we define backward-incompatible changes as changes that—when consumed—can potentially result in
errors for users that are using our public APIs or the entities that have `public` visibility in our code-base. >

- Any new external dependencies? **[YES/NO]**
  - < If YES, which ones and why? >
  - < In addition, please also mention any other alternatives you've considered and the reason they've been discarded >

- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **[YES/NO]**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by RCHowell** [comment]
See reference modeling as well as cast.unresolved vs cast.resolved. https://github.com/partiql/partiql-lang-kotlin/blob/partiql-spi-planner/partiql-planner/src/main/resources/partiql_plan_internal.ion/

File: partiql-plan/src/main/resources/partiql_plan.ion [hunk 1]
```diff
@@ -102,6 +114,11 @@ rex::{
       symbol::{ root: rex, key: string },
     ],
 
+    cast_op::{
+      arg: rex,
+      cast: cast
+    },
+
```
Comment (RCHowell) [inline partiql-plan/src/main/resources/partiql_plan.ion:121]: See modeling in https://github.com/partiql/partiql-lang-kotlin/blob/partiql-spi-planner/partiql-planner/src/main/resources/partiql_plan_internal.ion/

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/ir/Nodes.kt [hunk 2]
```diff
@@ -2,6 +2,7 @@
```
Comment (RCHowell) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/ir/Nodes.kt:2]: Consider rebasing so we have generate internal IRs

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/transforms/RexConverter.kt [hunk 3]
```diff
@@ -546,45 +555,10 @@ internal object RexConverter {
...
-                is Type.Interval -> TODO("Static Type does not have Interval type")
-                is Type.Bag -> rex(StaticType.BAG, call("cast_bag", arg0))
-                is Type.List -> rex(StaticType.LIST, call("cast_list", arg0))
-                is Type.Sexp -> rex(StaticType.SEXP, call("cast_sexp", arg0))
-                is Type.Tuple -> rex(StaticType.STRUCT, call("cast_tuple", arg0))
-                is Type.Struct -> rex(StaticType.STRUCT, call("cast_struct", arg0))
-                is Type.Any -> rex(StaticType.ANY, call("cast_any", arg0))
-                is Type.Custom -> TODO("Custom type not supported ")
-            }
+            val argType = arg0.type.toRuntimeType()
```
Comment (RCHowell) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/transforms/RexConverter.kt:560]: We have to use a resolved/unresolved cast variant. We don't know the type of the arg yet so we can't insert a cast. See https://github.com/partiql/partiql-lang-kotlin/blob/partiql-spi-planner/partiql-planner/src/main/kotlin/org/partiql/planner/internal/transforms/RexConverter.kt#L590

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/FnRegistry.kt [hunk 4]
```diff
@@ -76,11 +76,18 @@ internal class FnRegistry(private val metadata: Collection<ConnectorFunctions>)
...
+//        return if (rel.castType == CastType.COERCION) rel.castFn else null
+//    }
+
+    internal fun lookupCoercion(operand: PartiQLValueType, target: PartiQLValueType): TypeRelationship? {
         val i = operand.ordinal
         val j = target.ordinal
         val rel = pCasts.graph[i][j] ?: return null
-        return if (rel.castType == CastType.COERCION) rel.castFn else null
+        return if (rel.castType == CastType.COERCION) rel else null
     }
```
Comment (RCHowell) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/FnRegistry.kt:92]: This all goes away in https://github.com/partiql/partiql-lang-kotlin/blob/partiql-spi-planner

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/TypeCasts.kt [hunk 5]
```diff
@@ -103,9 +116,29 @@ internal class TypeCasts private constructor(
...
+            }
+        }
+
         /**
          * Build the PartiQL type lattice.
          *
+         * graph[operand] = [operand].relationship {
+         *    func(target)
+         * }
+         *
```
Comment (RCHowell) [inline partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/TypeCasts.kt:141]: Thanks for the clarification

**Review by RCHowell — CHANGES_REQUESTED** [review]
Left some references. I think these changes will need to be rebased on partiql-spi branch once planner updates are merged there

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/ir/Nodes.kt (hunk 2 above)
Comment (yliuuuu) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/ir/Nodes.kt:2]: I am thinking: 
1. partiql-spi-planner -> partiql-spi
2. partiql-plugin-impl -> partiql-spi
3. delete all partiql-plugin-xxxx
4. branch off partiql-spi to continue development
5. partiql-spi -> main

-> means get merged into.

**Comment by RCHowell** [comment]
I believe this is ready to be rebased on partiql-eval!

**Comment by yliuuuu** [comment]
close in favor of #1363

**Pull request closed without merging**